src/transformers/generation/utils.py

```diff
@@ -1517,7 +1517,7 @@ def generate(
         input_ids_length = input_ids.shape[-1]
         has_default_max_length = kwargs.get("max_length") is None and generation_config.max_length is not None
         if generation_config.max_new_tokens is not None:
-            if not has_default_max_length:
+            if not has_default_max_length or generation_config.max_length is not None:
```

Comment:
The logic in line could be simplified to `if kwargs.get("max_length") is not None:`, but I believe this way is much easier to understand the intent behind the warning :)